File: kotlin/robot-simulator/src/main/kotlin/RobotSimulator.kt
class Robot(var gridPosition: GridPosition = GridPosition(0, 0), var orientation: Orientation = Orientation.NORTH) {

    fun advance() {
        gridPosition = when (orientation) {
            Orientation.NORTH -> GridPosition(gridPosition.x, gridPosition.y + 1)
            Orientation.EAST -> GridPosition(gridPosition.x + 1, gridPosition.y)
            Orientation.SOUTH -> GridPosition(gridPosition.x, gridPosition.y - 1)
            Orientation.WEST -> GridPosition(gridPosition.x - 1, gridPosition.y)
            else -> GridPosition(0, 0)
        }
    }

    fun turnLeft() {
        orientation = Orientation.values()[(orientation.ordinal - 1 + 4) % 4]
    }

    fun turnRight() {
        orientation = Orientation.values()[(orientation.ordinal + 1) % 4]
    }

    fun simulate(instructions: String) {
        for (i in instructions) {
            when (i) {
                'A' -> advance()
                'L' -> turnLeft()
                'R' -> turnRight()
            }
        }
    }
}

File: dart/robot-simulator/lib/robot_simulator.dart
import 'package:robot_simulator/position.dart';
import 'package:robot_simulator/orientation.dart';

class Robot {
  Position position;
  Orientation orientation;

  Robot(Position p, Orientation o) {
    this.position = p;
    this.orientation = o;
  }

  void move(String command) {
    command.split('').forEach((c) {
      if (c == 'R') {
        orientation = Orientation.values[(orientation.index + 1) % 4];
      } else if (c == 'L') {
        orientation = Orientation.values[(orientation.index - 1 + 4) % 4];
      } else if (orientation == Orientation.north) {
        position.y++;
      } else if (orientation == Orientation.east) {
        position.x++;
      } else if (orientation == Orientation.south) {
        position.y--;
      } else {
        position.x--;
      }
    });
  }
}
